inference-engine/src/cldnn_engine/ops/loop.cpp

```diff
@@ -0,0 +1,250 @@
+        }
+    }
+
+    // get body topology from ngraph function
+    Program body_program(body_network, p.GetEnginePtr(), p.GetConfig(), true);
+    auto body_topology = *body_program.GetTopology();
+
+    // setup input_primitive_maps/ output_primitive_maps and back_edges
+    std::vector<cldnn::loop::io_primitive_map> input_primitive_maps;
+    std::vector<cldnn::loop::io_primitive_map> output_primitive_maps;
+    std::vector<cldnn::loop::backedge_mapping> back_edges;
+    std::map<cldnn::primitive_id, cldnn::primitive_id> reordered_output_ids;
+
+    // set input mapping & back edges
+    for (const auto& loop_input_desc : loop_input_descs) {
+        const cldnn::primitive_id& external_id = inputPrimitives.at(loop_input_desc->m_input_index);
+        auto& body_input = body_inputs.at(loop_input_desc->m_body_parameter_index);
+        cldnn::primitive_id internal_id = layer_type_name_ID(body_input);
+
+        // set input mapping
+        if (const auto& sliceInfo =
+            std::dynamic_pointer_cast<Loop::SliceInputDescription>(loop_input_desc)) {
+            // sliced input
+            input_primitive_maps.emplace_back(external_id, internal_id, sliceInfo->m_axis,
+                sliceInfo->m_start, sliceInfo->m_end, sliceInfo->m_stride);
+        } else {
+            // input without slicing
+            input_primitive_maps.emplace_back(external_id, internal_id);
+        }
+
+        // set back edges
+        if (const auto& mergedInput =
+            std::dynamic_pointer_cast<Loop::MergedInputDescription>(loop_input_desc)) {
+            // backedge
+            const auto& to = body_inputs.at(mergedInput->m_body_parameter_index);
+            const auto& from = body_outputs.at(mergedInput->m_body_value_index);
+
+            cldnn::primitive_id to_id = layer_type_name_ID(to);
+            cldnn::primitive_id from_id = layer_type_name_ID(from);
+
+            // reset output data type because the data types of the outputs of the
+            // body topology are always FP32 regardless of ngraph data type
+            {
+                const auto from_prim = body_topology.at(from_id);
+                const auto& to_ngraph_type = to->get_element_type();
+                const auto to_cldnn_type = DataTypeFromPrecision(to_ngraph_type);
+                from_prim->output_data_type = to_cldnn_type;
+            }
+            back_edges.emplace_back(from_id, to_id);
+        }
+    }
+
+    // set trip count, initial execution condition, num iteration primitives
+    // they should be mutable_data to prevent from being optimized out
+    const cldnn::primitive_id trip_count_id = layer_type_name_ID(op->get_input_node_shared_ptr(0));
+    const cldnn::primitive_id execution_condition_id = layer_type_name_ID(op->get_input_node_shared_ptr(1));
+    const int64_t num_iterations = op->get_num_iterations();
+    if (num_iterations < 0) {
+        throw std::runtime_error("loop's num_iteration cannot be negative");
```

Comment:
Please use `IE_THROW()` when possible (and preferably with explicit exception type)

Reply:
Fixed it. thanks.